File: app/src/main/java/com/m3sv/plainupnp/upnp/localcontent/AudioContainer.java
/**
 * Copyright (C) 2013 Aurélien Chabot <aurelien@chabot.fr>
 * <p>
 * This file is part of DroidUPNP.
 * <p>
 * DroidUPNP is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * <p>
 * DroidUPNP is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with DroidUPNP.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.m3sv.plainupnp.upnp.localcontent;

import android.content.Context;
import android.database.Cursor;
import android.provider.MediaStore;
import android.util.Log;

import org.droidupnp.legacy.mediaserver.ContentDirectoryService;
import org.fourthline.cling.support.model.PersonWithRole;
import org.fourthline.cling.support.model.Res;
import org.fourthline.cling.support.model.container.Container;
import org.fourthline.cling.support.model.item.MusicTrack;
import org.seamless.util.MimeType;

import java.util.List;

public class AudioContainer extends DynamicContainer {
    private static final String TAG = "AudioContainer";

    public AudioContainer(String id, String parentID, String title, String creator, String baseURL, Context ctx,
                          String artist, String albumId) {
        super(id, parentID, title, creator, baseURL, ctx, null);
        setUri(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI);

        if (artist != null) {
            setWhere(MediaStore.Audio.Media.ARTIST + "=?");
            setWhereVal(new String[]{artist});
            setOrderBy(MediaStore.Audio.Media.ALBUM);
        }

        if (albumId != null) {
            setWhere(MediaStore.Audio.Media.ALBUM_ID + "=?");
            setWhereVal(new String[]{albumId});
            setOrderBy(MediaStore.Audio.Media.TRACK);
        }
    }

    @Override
    public Integer getChildCount() {
        String[] columns = {MediaStore.Audio.Media._ID};
        Cursor cursor = getCtx().getContentResolver().query(getUri(), columns, getWhere(), getWhereVal(), getOrderBy());
        if (cursor == null)
            return 0;

        int result = cursor.getCount();
        cursor.close();
        return result;
    }

    @Override
    public List<Container> getContainers() {
        String[] columns = {
                MediaStore.Audio.Media._ID,
                MediaStore.Audio.Media.TITLE,
                MediaStore.Audio.Media.DATA,
                MediaStore.Audio.Media.ARTIST,
                MediaStore.Audio.Media.MIME_TYPE,
                MediaStore.Audio.Media.SIZE,
                MediaStore.Audio.Media.DURATION,
                MediaStore.Audio.Media.ALBUM
        };

        Cursor cursor = getCtx().getContentResolver().query(getUri(), columns, getWhere(), getWhereVal(), getOrderBy());
        if (cursor != null) {
            if (cursor.moveToFirst()) {
                do {
                    String id = ContentDirectoryService.AUDIO_PREFIX + cursor.getInt(cursor.getColumnIndex(MediaStore.Audio.Media._ID));
                    String title = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE));
                    String creator = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST));
                    String filePath = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DATA));
                    String mimeType = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.MIME_TYPE));
                    long size = cursor.getLong(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.SIZE));
                    long duration = cursor.getLong(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DURATION));
                    String album = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM));

                    String extension = "";
                    int dot = filePath.lastIndexOf('.');
                    if (dot >= 0)
                        extension = filePath.substring(dot).toLowerCase();

                    Res res = new Res(new MimeType(mimeType.substring(0, mimeType.indexOf('/')),
                            mimeType.substring(mimeType.indexOf('/') + 1)), size, "http://" + getBaseURL() + "/" + id + extension);

                    res.setDuration(duration / (1000 * 60 * 60) + ":"
                            + (duration % (1000 * 60 * 60)) / (1000 * 60) + ":"
                            + (duration % (1000 * 60)) / 1000);

                    addItem(new MusicTrack(id, parentID, title, creator, album, new PersonWithRole(creator, "Performer"), res));

                    Log.v(TAG, "Added audio item " + title + " from " + filePath);

                } while (cursor.moveToNext());
            }
            cursor.close();
        }

        return containers;
    }

}

File: upnp/src/main/kotlin/com/m3sv/plainupnp/upnp/mediacontainers/AudioContainer.kt
package com.m3sv.plainupnp.upnp.mediacontainers

import android.content.ContentResolver
import android.os.Build
import android.provider.MediaStore
import com.m3sv.plainupnp.upnp.ContentDirectoryService
import org.fourthline.cling.support.model.PersonWithRole
import org.fourthline.cling.support.model.Res
import org.fourthline.cling.support.model.container.Container
import org.fourthline.cling.support.model.item.MusicTrack
import org.seamless.util.MimeType

class AudioContainer(
    id: String,
    parentID: String,
    title: String,
    creator: String?,
    baseURL: String,
    private val contentResolver: ContentResolver,
    private val directory: String?,
    artist: String? = null,
    albumId: String? = null
) : DynamicContainer(id, parentID, title, creator, baseURL) {

    private var orderBy: String? = null
    private var where: String? = null
    private var whereVal: Array<String>? = null

    private val selection: String? = if (directory != null) "$AUDIO_DATA_PATH LIKE ?" else null

    private val selectionArgs: Array<String>? =
        if (directory != null) arrayOf("%$directory/") else null

    init {
        val directoryClause = if (selection != null) {
            " ,$selection"
        } else
            ""

        if (directory != null) {
            where = selection
            whereVal = selectionArgs
        }

        if (artist != null) {
            where = "${MediaStore.Audio.Media.ARTIST} = ?${directoryClause}"
            whereVal = if (directory != null) {
                arrayOf(artist, "%$directory/")
            } else
                arrayOf(artist)
            orderBy = MediaStore.Audio.Media.ALBUM
        }

        if (albumId != null) {
            where = "${MediaStore.Audio.Media.ALBUM_ID} = ?${directoryClause}"
            whereVal = if (directory != null) {
                arrayOf(albumId, "%$directory/")
            } else
                arrayOf(albumId)

            orderBy = MediaStore.Audio.Media.TRACK
        }
    }

    override fun getChildCount(): Int? {
        val projection = if (directory != null)
            arrayOf(
                MediaStore.Audio.Media._ID,
                AUDIO_DATA_PATH
            )
        else
            arrayOf(MediaStore.Audio.Media._ID)

        contentResolver.query(
            URI,
            projection,
            where,
            whereVal,
            orderBy
        ).use { cursor ->
            return cursor?.count ?: 0
        }
    }

    override fun getContainers(): List<Container> {
        if (items.isNotEmpty() || containers.isNotEmpty())
            return containers

        val columns = arrayOf(
            MediaStore.Audio.Media._ID,
            MediaStore.Audio.Media.TITLE,
            MediaStore.Audio.Media.ARTIST,
            MediaStore.Audio.Media.MIME_TYPE,
            MediaStore.Audio.Media.SIZE,
            MediaStore.Audio.Media.DURATION,
            MediaStore.Audio.Media.ALBUM
        )

        contentResolver.query(
            URI,
            columns,
            where,
            whereVal,
            orderBy
        )?.use { cursor ->
            val audioIdColumn = cursor.getColumnIndex(MediaStore.Audio.Media._ID)
            val audioTitleColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.TITLE)
            val audioArtistColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ARTIST)
            val audioMimeTypeColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.MIME_TYPE)
            val audioSizeColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.SIZE)
            val audioDurationColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.DURATION)
            val audioAlbumColumn = cursor.getColumnIndexOrThrow(MediaStore.Audio.Media.ALBUM)

            while (cursor.moveToNext()) {
                val id = ContentDirectoryService.AUDIO_PREFIX + cursor.getInt(audioIdColumn)
                val title = cursor.getString(audioTitleColumn)
                val creator = cursor.getString(audioArtistColumn)
                val type = cursor.getString(audioMimeTypeColumn)
                val size = cursor.getLong(audioSizeColumn)
                val duration = cursor.getLong(audioDurationColumn)
                val album = cursor.getString(audioAlbumColumn)

                val mimeType = type.substring(0, type.indexOf('/'))
                val mimeSubType = type.substring(type.indexOf('/') + 1)
                val res = Res(
                    MimeType(mimeType, mimeSubType),
                    size,
                    "http://$baseUrl/$id.$mimeSubType"
                )

                res.duration =
                    "${(duration / (1000 * 60 * 60))}:${duration % (1000 * 60 * 60) / (1000 * 60)}:${duration % (1000 * 60) / 1000}"

                addItem(
                    MusicTrack(
                        id,
                        parentID,
                        title,
                        creator,
                        album,
                        PersonWithRole(creator, "Performer"),
                        res
                    )
                )
            }
        }

        return containers
    }

    companion object {
        private val URI = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI

        val AUDIO_DATA_PATH = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
            MediaStore.Audio.Media.RELATIVE_PATH
        else
            MediaStore.Audio.Media.DATA
    }
}
